core/src/SYCL/Kokkos_SYCL_Parallel_Reduce.hpp

```diff
@@ -404,42 +403,60 @@ class ParallelReduce<CombinedFunctorReducerType, Kokkos::RangePolicy<Traits...>,
         auto multiple = kernel.get_info<sycl::info::kernel_device_specific::
                                             preferred_work_group_size_multiple>(
             q.get_device());
+        // FIXME_SYCL The code below queries the kernel for the maximum subgroup
+        // size but it turns out that this is not accurate and choosing a larger
+        // subgroup size gives better peformance (and is what the SYCL compiler
+        // does for sycl::reductions).
+#ifndef KOKKOS_ARCH_INTEL_GPU
         auto max =
             kernel
                 .get_info<sycl::info::kernel_device_specific::work_group_size>(
                     q.get_device());
-
-// FIXME_SYCL 1024 seems to be invalid when running on a Volta70.
-#ifndef KOKKOS_ARCH_INTEL_GPU
-        if (max > 512) max = 512;
+#else
+        auto max =
+            q.get_device().get_info<sycl::info::device::max_work_group_size>();
 #endif
 
-        const size_t wgroup_size =
-            static_cast<size_t>(max / multiple) * multiple;
+        auto max_local_memory =
+            q.get_device().get_info<sycl::info::device::local_mem_size>();
+        // Use a maximum of 99% of the available memory as a safe-guard.
+        const auto wgroup_size = std::min(
+            {Kokkos::bit_ceil(size),
+             static_cast<size_t>(max / multiple) * multiple,
+             Kokkos::bit_floor(static_cast<size_t>(max_local_memory * .99) /
+                               (sizeof(value_type) * value_count))});
```

Comment:
Please comment about this (in code)
Are you are trying to get a power of two?

Reply:
9ac49e9